runhouse/rns/run_module_utils.py

```diff
@@ -14,80 +16,112 @@
+                for arg in args
+            ]
+        elif fn_type in ("queue", "remote", "call", "nested"):
+            obj_ref = ray_fn.remote(
+                fn_pointers, fn_type, num_cuda_devices, *args, **kwargs
+            )
         elif fn_type == "repeat":
             [num_repeats, args] = args
-            obj_ref = [ray_fn.remote(*args, **kwargs) for _ in range(num_repeats)]
+            obj_ref = [
+                ray_fn.remote(fn_pointers, fn_type, num_cuda_devices, *args, **kwargs)
+                for _ in range(num_repeats)
+            ]
         else:
             raise ValueError(f"fn_type {fn_type} not recognized")
 
         if fn_type == "remote":
             rh_config.obj_store.put_obj_ref(key=run_key, obj_ref=obj_ref)
-            res = run_key
-        else:
+            res = pickle.dumps(run_key)
+        elif fn_type in ("call", "nested"):
             res = ray.get(obj_ref)
+        else:
+            res = pickle.dumps(ray.get(obj_ref))
     return res
 
 
-def get_fn_by_name(module_name, fn_name, relative_path=None):
-    if relative_path:
-        module_path = str((Path.home() / relative_path).resolve())
-        sys.path.append(module_path)
-        logger.info(f"Appending {module_path} to sys.path")
-
-    if module_name in rh_config.obj_store.imported_modules:
-        importlib.invalidate_caches()
-        rh_config.obj_store.imported_modules[module_name] = importlib.reload(
-            rh_config.obj_store.imported_modules[module_name]
-        )
-        logger.info(f"Reloaded module {module_name}")
+def get_fn_from_pointers(fn_pointers, fn_type, num_gpus, *args, **kwargs):
+    (module_path, module_name, fn_name) = fn_pointers
+    if module_name == "notebook":
+        fn = fn_name  # already unpickled
     else:
-        logger.info(f"Importing module {module_name}")
-        rh_config.obj_store.imported_modules[module_name] = importlib.import_module(
-            module_name
-        )
-    fn = getattr(rh_config.obj_store.imported_modules[module_name], fn_name)
-    return fn
+        if module_path:
+            sys.path.append(module_path)
+            logger.info(f"Appending {module_path} to sys.path")
+
+        if module_name in rh_config.obj_store.imported_modules:
+            importlib.invalidate_caches()
+            rh_config.obj_store.imported_modules[module_name] = importlib.reload(
+                rh_config.obj_store.imported_modules[module_name]
+            )
+            logger.info(f"Reloaded module {module_name}")
+        else:
+            logger.info(f"Importing module {module_name}")
+            rh_config.obj_store.imported_modules[module_name] = importlib.import_module(
+                module_name
+            )
+        fn = getattr(rh_config.obj_store.imported_modules[module_name], fn_name)
+
+    cuda_visible_devices = list(range(int(num_gpus)))
+    os.environ["CUDA_VISIBLE_DEVICES"] = ",".join(map(str, cuda_visible_devices))
```

Comment:
replace this with setting `export RAY_EXPERIMENTAL_NOSET_CUDA_VISIBLE_DEVICES=1` prior to ray init

Reply:
tried setting `os.environ["RAY_EXPERIMENTAL_NOSET_CUDA_VISIBLE_DEVICES"] = "1"` prior to the `ray.init` in unary_server, but looked like it was still blocking cuda and I would only see 1 gpu within the function call